File: Java test/Test/src/Complex.java
public interface Complex {
    public double realpart ( );        // returns this.real
    public double imagpart( );         // returns this.imag
    public double angle ( );           // returns this.angle
    public double magnitude      ( );  // returns this.mag
    public void  add   (Complex c);    // this = this + c
    public void  minus (Complex c);    // this = this - c
    public void  times (Complex c);    // this = this * c
    public void  divides(Complex c);
    public Complex conjugate();

    public static void main(String[] args){
        Complex c3plus4i = new ComplexCart(3,4);
        Complex c5plus6i = new ComplexCart(5,6);
        c3plus4i.divides(c5plus6i);
        System.out.println(c3plus4i);
    }
}

class ComplexCart implements Complex {
    private double real;
    private double imag;
    // CONSTRUCTORS
    public ComplexCart (double r, double i) { real = r; imag = i; }

    // ACCESSORS
    public double realpart ()  {  return this.real; }
    public double imagpart ()  { return this.imag; }
    public double magnitude () { return Math.sqrt(real * real + imag * imag); }
    public double angle ()  {
        if (real != 0) {
            if (real < 0 ) return (Math.PI + Math.atan(imag/real));
            return Math.atan(imag / real);
        }
        else if (imag == 0) return 0;
        else if (imag > 0) return Math.PI/2;
        else return -Math.PI/2;
    }

    // MUTATORS
    public void add (Complex c)  {
        this.real += c.realpart();
        this.imag += c.imagpart();
    }

    public void minus (Complex c)   {
        this.real -= c.realpart();
        this.imag -= c.imagpart();
    }

    public void times (Complex c)   {
        double tempReal=real*c.realpart() - imag*c.imagpart();
        imag = real * c.imagpart() + imag * c.realpart();
        real = tempReal;
    }

    public void divides (Complex c) {
        // Write your code here
        Complex denom = new ComplexCart(c.realpart(), c.imagpart());
        denom.times(denom.conjugate()); // which should equal real^2 + imag^2

        this.times(c.conjugate());

        //denom is a real number, hence only taking the real part
        //ie denom = a + bi, where b = 0
        real /= denom.realpart();
        imag /= denom.realpart();
    }

    public Complex conjugate (){
        return new ComplexCart(real, -imag);
    }

    // for printing...
    public String toString() {
        if ( imag == 0 ) return ( realpart() + "");
        else if ( imag < 0 )
            return ( real + "" + imag + "i");
        else
            return ( real + "+" + imag + "i");
    }
}

class ComplexPolar implements Complex {
    private double mag;   // magnitude
    private double ang;   // angle
    // CONSTRUCTORS
    ComplexPolar (double m, double r) { mag = m; ang = r; }

    // ACCESSORS
    public double realpart ()  { return mag * Math.cos(ang); }
    public double imagpart ()  { return mag * Math.sin(ang); }
    public double magnitude () { return mag; }
    public double angle ()     { return ang; }

    // MUTATORS
    public void add (Complex c)   {     // this = this + c
        double real =  realpart() + c.realpart();
        double imag =  imagpart() + c.imagpart();
        mag = Math.sqrt(real*real + imag*imag);
        if (real != 0) {
            if (real < 0) ang = (Math.PI + Math.atan(imag/real));
            else ang = Math.atan(imag/real);
        }
        else if (imag == 0) ang = 0;
        else if(imag > 0) ang = Math.PI/2;
        else ang = -Math.PI/2;
    }

    public void minus (Complex c)   {     // this = this - c
        double real =  mag * Math.cos(ang) - c.realpart();
        double imag =  mag * Math.sin(ang) - c.imagpart();
        mag = Math.sqrt(real*real + imag*imag);
        if (real != 0) {
            if (real < 0) ang = (Math.PI + Math.atan(imag/real));
            else ang = Math.atan(imag/real);
        }
        else if (imag == 0) ang = 0;
        else if(imag > 0) ang = Math.PI/2;
        else ang = -Math.PI/2;
    }

    public void times (Complex c)  {   // this = this * c
        mag *= c.magnitude();
        ang += c.angle();
        ang = ang % (2*Math.PI); // maintain ang within 2*pi
    }

    public void divides (Complex c) {
        // Write your code here
        mag /= c.magnitude();
        ang -= c.angle();
    }

    public Complex conjugate(){
        return new ComplexPolar(mag, -ang);
    }

    public String toString() {
        if ( imagpart() == 0 ) return ( this.realpart() + "");
        else if ( imagpart() < 0 )
            return ( realpart() + "" + imagpart() + "i");
        else
            return ( realpart() + "+" + imagpart() + "i");
    }
}

File: Week 17/Lecture15/Complex Number/Complex.py
from math import *

class Complex (object):
    def __init__(self):
        pass
    def realpart(self):
        pass
    def imagpart(self):
        pass
    def angle(self):
        pass
    def mag(self):
        pass
    def add(self, c):
        pass
    def minus(self, c):
        pass
    def times(self, c):
        pass
    def toString(self):
        if (self.imagpart() == 0):
            return str(self.realpart())
        elif (self.imagpart() < 0):
            return str(self.realpart())+str(self.imagpart())+"i"
        else:
            return str(self.realpart())+"+"+str(self.imagpart())+"i"


class ComplexCart(Complex):
    def __init__(self, r, i):
        self.real = r
        self.imag = i
    def realpart(self):
        return self.real
    def imagpart(self):
        return self.imag
    def magnitude(self):
        return sqrt(self.realpart()*self.realpart()+self.imag*self.imag)   
    def angle(self):
        if self.real!=0:
            if (self.real < 0):
                return pi + atan(self.imag/self.real)
            return atan(self.imag/self.real)
        elif self.imag == 0:
            return 0
        elif self.imag > 0:
            return pi/2
        else:
            return -pi/2     
    def add(self, c):
        self.real += c.realpart()
        self.imag += c.imagpart()
    def minus(self, c):
        self.real -= c.realpart()
        self.imag -= c.imagpart()
    def times(self, c):
        tempReal = self.real * c.realpart() - self.imag * c.imagpart()
        self.imag = self.real * c.imagpart() + self.imag * c.realpart()
        self.real = tempReal

class ComplexPolar(Complex):
    def __init__(self, m, a):
        self.mag = m
        self.ang = a
    def realpart(self):
        return self.mag * cos(self.ang)
    def imagpart(self):
        return self.mag * sin(self.ang)
    def magnitude(self):
        return self.mag
    def angle(self):
        return self.ang   
    def add(self, c):
        real = self.realpart() + c.realpart()
        imag = self.imagpart() + c.imagpart()
        self.mag = sqrt(real*real + imag*imag)
        if (real != 0):
            if (real < 0):
                self.ang = pi + atan(imag/real)
            else:
                self.ang = atan(imag/real)
        elif (imag == 0):
            self.ang = 0
        elif (imag > 0):
            self.ang = pi/2
        else:
            self.ang = -pi/2         
    def minus(self, c):
        real = self.realpart() - c.realpart()
        imag = self.imagpart() - c.imagpart()
        self.mag = sqrt(real*real + imag*imag)
        if (real != 0):
            if (real < 0):
                self.ang = pi + atan(imag/real)
            else:
                self.ang = atan(imag/real)
        elif (imag == 0):
            self.ang = 0
        elif (imag > 0):
            self.ang = pi/2
        else:
            self.ang = -pi/2         
    def times(self, c):
        self.mag *= c.magnitude()
        self.ang += c.angle()
        self.ang = self.ang % (2*pi)


# testing ComplexCart
a = ComplexCart(10.0, 12.0)
b = ComplexCart(1.0, 2.0)
a.add(b)
print("a=a+b is", a.toString())
a.minus(b)
print("a-b (which is the original a) is", a.toString())
print("Angle of a is", str(a.angle()))
a.times(b)
print("a x b is", a.toString())

# Testing ComplexPolar
c = ComplexPolar(10.0,pi/6.0)
d = ComplexPolar(10.0, pi/3.0)
print("c is",c.toString())
print("d is", d.toString())
c.add(d)
print("c=c+d is", c.toString())
c.minus(d)
print("c-d (which is the original c) is ", c.toString())
c.times(d)
print("c=c*d is", c.toString())

# Testing Combine
print("a is", a.toString())
print("d is", d.toString())
a.minus(d)
print("a=a-d is", a.toString()) 
a.times(d)
print("a=a*d is", a.toString())
d.add(a)
print("d=d+a is", d.toString())
d.times(a)
print("d=d*a is", d.toString()) 


